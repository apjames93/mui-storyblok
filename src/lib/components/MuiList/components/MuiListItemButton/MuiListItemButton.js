import React from 'react';
import PropTypes from 'prop-types';
import ListItem from '@material-ui/core/ListItem';
import { validComponents, muiStringProp } from '../../../../utils/customProps';
import Storyblok from '../../../../utils/Storyblok';
import appRedirect from '../../../../utils/appRedirect';
import MuiListItemAvatar from '../MuiListItem/components/MuiListItemAvatar/MuiListItemAvatar';
import MuiListItemIcon from '../MuiListItem/components/MuiListItemIcon/MuiListItemIcon';
import MuiListItemText from '../MuiListItem/components/MuiListItemText/MuiListItemText';
import MuiContactButton from '../../../MuiContactButton/MuiContactButton';

export const MuiListItemButton = ({
  rootClass,
  alignItems,
  dense,
  disableGutters,
  divider,
  selected,
  redirectRoute,
  contactButton,
  href,
  listItemAvatar,
  listItemIcon,
  listItemText,
}) => {
  const styles = Storyblok.arrayToMuiStyles(rootClass);
  let handleClick;
  if (href === undefined && redirectRoute === undefined) {
    handleClick = () => {};
  } else if (href !== undefined && href !== '') {
    handleClick = () => window.location.assign(href);
  } else if (redirectRoute !== undefined && redirectRoute !== '') {
    handleClick = async () => appRedirect(redirectRoute);
  }

  const avatar = listItemAvatar[0];
  const icon = listItemIcon[0];
  const text = listItemText[0];
  const contact = contactButton[0];
  return (
    <ListItem
      className={styles.root}
      dense={dense}
      alignItems={alignItems}
      disableGutters={disableGutters}
      divider={divider}
      selected={selected}
      button={true}
      onClick={handleClick}
    >
      {avatar ? <MuiListItemAvatar {...avatar} /> : null}
      {icon ? <MuiListItemIcon {...icon} /> : null}
      {text ? <MuiListItemText {...text} /> : null}
      {contact ? <MuiContactButton {...contact} /> : null}
    </ListItem>
  );
};

export default MuiListItemButton;

MuiListItemButton.propTypes = {
  /** storyblok multiselect of css classes */
  rootClass: PropTypes.arrayOf(PropTypes.string),
  /** mui prop: 'flex-start'| 'center'
   * Defines the align-items style property.
  */
  alignItems(props, propName, componentName) {
    return muiStringProp(props, propName, componentName, ['flex-start', 'center']);
  },
  /** mui prop: true | false
   * If true, compact vertical padding designed for keyboard and mouse input will be used.
   * */
  dense: PropTypes.bool,
  /** mui prop: true | false
   * If true, the left and right padding is removed.
   * */
  disableGutters: PropTypes.bool,
  /** mui prop: true | false
   * If true, a 1px light border is added to the bottom of the list item.
   * */
  divider: PropTypes.bool,
  /** mui prop: true | false
   * Use to apply selected styling.
   * */
  selected: PropTypes.bool,
  /** redirect route */
  redirectRoute: PropTypes.string,
  /** url to redirect to */
  href: PropTypes.string,
  /** MuiListItemAvatar Allowed maximum: 1 */
  contactButton(props, propName, componentName) {
    const components = ['MuiListItemAvatar'];
    return validComponents(props, propName, componentName, components, 1);
  },
  /** MuiListItemAvatar Allowed maximum: 1 */
  listItemAvatar(props, propName, componentName) {
    return validComponents(props, propName, componentName, ['MuiListItemAvatar'], 1);
  },
  /** MuiListItemIcon Allowed maximum: 1 */
  listItemIcon(props, propName, componentName) {
    return validComponents(props, propName, componentName, ['MuiListItemIcon'], 1);
  },
  /** MuiListItemText Allowed maximum: 1 */
  listItemText(props, propName, componentName) {
    return validComponents(props, propName, componentName, ['MuiListItemText'], 1);
  },
};

MuiListItemButton.defaultProps = {
  alignItems: 'center',
  dense: false,
  disableGutters: false,
  divider: false,
  selected: false,
  redirectRoute: undefined,
  href: undefined,
  rootClass: [],
  listItemAvatar: [],
  listItemIcon: [],
  listItemText: [],
  contactButton: [],
};
