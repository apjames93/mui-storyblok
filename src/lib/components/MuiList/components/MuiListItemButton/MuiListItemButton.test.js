import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { MuiListItemButton } from './MuiListItemButton';

function setup(isButton = false, redirectRoute = undefined, href = undefined) {
  const props = {
    _editable: '<!--#storyblok#{"name": "MuiListItemButton", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',

    listItemAvatar: [{
      _editable: '<!--#storyblok#{"name": "MuiListItemAvatar", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',

      component: 'MuiListItemAvatar',
      content: [{
        _editable: '<!--#storyblok#{"name": "MuiIcon", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',

        component: 'MuiIcon',
        iconName: 'android',
      }],
    }],

    listItemIcon: [{
      _editable: '<!--#storyblok#{"name": "MuiListItemIcon", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',

      component: 'MuiListItemIcon',
      content: [{
        _editable: '<!--#storyblok#{"name": "MuiIcon", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',

        component: 'MuiIcon',
        iconName: 'android',
      }],
    }],

    listItemText: [{
      _editable: '<!--#storyblok#{"name": "MuiListItemText", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',

      component: 'MuiListItemText',
      primary: 'primary',
    }],
    isButton,
    redirectRoute,
    history: {
      push: jest.fn(),
    },
    href,
  };
  const comp = shallow(<MuiListItemButton {...props} />);
  return { comp, props };
}

describe('<MuiListItemButton />', () => {
  it('renders MuiListItemButton', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  it('handleClick and calls history push ', async () => {
    const { comp } = setup(true, '/page-test');
    window.location.assign = jest.fn();
    comp.find('WithStyles(ForwardRef(ListItem))').at(0).simulate('click');
    expect(window.location.assign).toBeCalled();
  });

  it('handleClick and calls history push ', async () => {
    const { assign } = window.location;
    window.location.assign = jest.fn();
    const { comp } = setup(false, undefined, 'www.google.com');
    comp.find('WithStyles(ForwardRef(ListItem))').at(0).simulate('click');
    expect(window.location.assign).toBeCalled();
    window.location.assign = assign;
  });

  test('snapshot', () => {
    const { props } = setup();
    const tree = renderer.create(<MuiListItemButton {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
