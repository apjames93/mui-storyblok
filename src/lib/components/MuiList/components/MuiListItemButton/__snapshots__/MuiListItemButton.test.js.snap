// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MuiListItemButton /> snapshot 1`] = `
<div
  aria-disabled={false}
  className="MuiButtonBase-root MuiListItem-root makeStyles-root-5 MuiListItem-gutters MuiListItem-button"
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex={0}
>
  <div
    className="MuiListItemAvatar-root makeStyles-root-6"
  >
    <div
      className="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
    >
      <span
        aria-hidden={true}
        className="material-icons MuiIcon-root makeStyles-root-7 MuiIcon-colorSecondary"
      >
        android
      </span>
    </div>
  </div>
  <div
    className="MuiListItemIcon-root makeStyles-root-8"
  >
    <span
      aria-hidden={true}
      className="material-icons MuiIcon-root makeStyles-root-9 MuiIcon-colorSecondary"
    >
      android
    </span>
  </div>
  <div
    className="MuiListItemText-root makeStyles-root-10 MuiListItemText-multiline"
  >
    <span
      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
    >
      primary
    </span>
    <p
      className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
    >
      
    </p>
  </div>
</div>
`;
